File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.cpp
class Solution {
public:
    string reverseWords(string s) {
        int index = 0;
        int n = s.size();
        int i = 0;
        while(i<n){
            while(i<n && s[i] == ' '){
                i++;
            }
            if(i>=n) break;
            if(index > 0){
                s[index++] = ' ';
            }

            while(index < n && i < n && s[i] != ' '){
                s[index++] = s[i++];
            }
        }
        n = index;
        s.resize(n);
        reverse(s.begin(),s.end());
        int start = 0;
        for(int j=0;j<n;j++){
            if(s[j] == ' '){
                reverse(s.begin() + start, s.begin() + j);
                start = j + 1;
            }
        }
        reverse(s.begin() + start, s.end());
        return s;
    }
};

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.java
class Solution {
    public String reverseWords(String s) {
        String res = "";
        String temp = "";
        for(int i=0;i<s.length();i++){
            if(s.charAt(i) == ' '){
                if(temp == "") continue;
                res = temp + (res == "" ? "" : " " + res);
                temp = "";
            }else{
                temp += s.charAt(i);
            }
        }
        if(temp != ""){
            res = temp + (res == "" ? "" : " " + res);
        }
        return res;
    }
}